### ISSUE TITLE
XML has not well formed (invalid token)

### ISSUE THREAD

**Issue by Symon3** [OCSInventory-NG/UnixAgent#275]
*It's hard to solve a problem when important details are missing, that why we added this template, to help you and us.*

### General informations
Operating system :  CentOS7
Perl version : 5.16.3

### OCS Inventory informations
Unix agent version : 2.6.0

### Problem's description

I installed OCSinventory agent (2.6.0) onto a CentOS7 machine. It's reporting to an OCS inventory server, version 2.7.0. It fails with this error in the agent logfile:-

"[error] Cannot establish communication : 500 Internal Server Error"

The server has this error in the httpd error file:-

"not well-formed (invalid token) at line 931, column 18"

So, I did a debug run of the OCS agent, and the data around line 931 in the agent's XML file is:-
```
    <MONITORS>
      <BASE64>AP///////wAe1gcJPTa9ABoNAQJoHhd46mP4nlZMmCQZUFW/74DRQIGAgYqBj6lAqUqpTwEBAAAA
/ABLVk0ARzU2MTAgCiAgAAAA/AAxOTIweDE0NDAKAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAOk=
</BASE64>
      <CAPTION>KVM^@G5610 1920x1440</CAPTION>
      <DESCRIPTION>26/2003</DESCRIPTION>
      <MANUFACTURER>Unknown manufacturer code GVV</MANUFACTURER>
      <SERIAL>12400189</SERIAL>
    </MONITORS>
```

I believe the error is because of the ^@ character in the CAPTION line. Any ideas how to fix this?

### Inventory log file ( optional )
[Wed Jun 24 12:23:15.984920 2020] [perl:error] [pid 6481] [client 192.168.38.12:46949] \nnot well-formed (invalid token) at line 931, column 18, byte 34540 at /usr/lib64/perl5/vendor_perl/XML/Parser.pm line 187.\n

**Comment by somzoli** [comment]
Hi

Did you find a solution for this? I have also this error.

Z

**Comment by Symon3** [comment]
Sadly not.

**Comment by fbomj** [comment]
Hi

Could you please post the characteristics of the screen, please ? Manufacturer etc

Regards

**Comment by Symon3** [comment]
OK, I went back through my Zabbix data. The machine is a PowerEdge R430 server. That machine has since been re-purposed. I might be able to log into it later, if I can track it down. However, I think it was headless, and someone used a KVM switch on it. I imagine we're most interested in that, rather than the server. From searching for the 5610 number in the logs, I strongly suspect it was an APC AP5610 KVM switch. I'll get back to you.
p.s. I think the issue is because the data isn't sanitized properly before being added to the XML file.

**Comment by somzoli** [comment]
For me:

I installed Unix agent 2.8.0 on macos. It works, but not reported hardware details (CPU,MEM,NETWORK). After that, i installed "Mac::Sysprofile" perl module, than i get 500 internal server error on client and not well-formed (invalid token) error on server side apache. 
So i think also the XML is not correctly formatted before send.

**Comment by Symon3** [comment]
So, I spoke to the guy at the place where the server is, and he says the KVM was a W119b, it looks like this:- https://www.kvm-switches-online.com/w119.html

**Comment by somzoli** [comment]
Hi,
I was make an stdout report to report.xml and then make an xml well fromatted test with command line tools:

perl -MXML::Parser -e "XML::Parser->new( ErrorContext => 3 )->parsefile(shift)" /tmp/report.xml

And then the result:

not well-formed (invalid token) at line 1300, column 19, byte 33777:
    <MEMORIES>
      <CAPACITY>4096</CAPACITY>
      <CAPTION>Status: ok</CAPTION>
      <DESCRIPTION></DESCRIPTION>
==================^
      <SERIALNUMBER>-</SERIALNUMBER>
      <SPEED>1867 MHz</SPEED>
      <TYPE>DDR3</TYPE>
 at /usr/local/Cellar/perl/5.32.1_1/lib/perl5/site_perl/5.32.1/darwin-thread-multi-2level/XML/Parser.pm line 187.


I think the problem is arround the Description on MacOs. 

Can you help with this?

Thank you,
Z

**Comment by fbomj** [comment]
@somzoli => apply this fix #339

**Comment by wiltonsr** [comment]
This issue is a little more complicated. I had the same problem with an LG monitor which upon being invented returns a non-ASCII character which produces the same problem.

In my case this error occurred in the module `/Ocsinventory/Agent/Backend/OS/Generic/Screen.pm` when inventorying the `Caption` information, which resulted in the XML being generated like this:

`<CAPTION>L17F^A</CAPTION>`

Which consequently also resulted in the not well-formed (invalid token) error being processed by the server.

As seen in the case of @Symon3 and mine, non-ASCII characters are not allowed in XML fields, that is, this error can occur in any of the inventoried information if the information value has invalid characters.

And in the case of @somzoli we verified that it is not possible to have empty tags either.

I was able to fix my problem temporarily by adding the following line to the Screen.pm module:

`$caption =~ tr/\x20-\x7f//cd;`

However, I believe that performing an XML check on the agent before it is sent to the server to validate that there are no invalid characters or empty fields would be a more robust way to try to prevent this type of error from occurring again.

**Comment by thfiliol1** [comment]
Hi, I have the same problem! so to resolve it: 
You can also use this code: 
if (!exists $found{$serial}) {
            $found{$serial} = 1;
 
            eval "use MIME::Base64;";
            $base64 = encode_base64($raw_edid) if !$@;
            if ($common->can_run("uuencode")) {
                chomp($uuencode = `echo $raw_edid|uuencode -`);
                if (!$base64) {
                    chomp($base64 = `echo $raw_edid|uuencode -m -`);
                }
            }
            **$caption =~ s/[^a-zA-Z0-9]//g;**
            $common->addMonitor ({
                BASE64 => $base64,
                CAPTION => $caption,
                DESCRIPTION => $description,
                MANUFACTURER => $manufacturer,
                SERIAL => $serial,
                UUENCODE => $uuencode,
            });
        }
    }

However I share the same idea that @wiltonsr to check and delete all special characters for all XML data which is send to the OCS Server

**Comment by fbomj** [comment]
Fix with PR #360
